### PR TITLE
feat(outputs.parquet): Introduce Parquet output

### PR DESCRIPTION
**Pull request by powersj**
## Summary

Introduces a new output to write metrics in Parquet format. This groups metrics by metric name and writes them to files. We need to know the schema to write beforehand, the first time we encounter a metric name we generate the schema based on the metric we found then use the buffered file writer from here on out to write files as efficiently as possible.

We must close the file correctly otherwise the file will not be a valid parquet file. As such I have avoided creating files by template name due to complications with keeping too many files open. Instead, I have a time-based rotation option that will close the existing file and create a new file.

Check out this blog post for an overview of Parquet files: https://www.influxdata.com/blog/how-good-parquet-wide-tables/

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

fixes: #14786

### CODE REVIEW

File: plugins/outputs/parquet/README.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# Parquet Output Plugin
+
+This plugin sends writes metrics to parquet files. By default, the parquet
+output will groups metrics by metric name and write those metrics all to the
```
Comment (DStrand1) [inline plugins/outputs/parquet/README.md:4]:
```suggestion
This plugin writes metrics to parquet files. By default, the parquet
output will group metrics by metric name and write those metrics all to the
```

File: plugins/outputs/parquet/README.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+rotated to avoid over-writing it or conflicting schema.
+
+File rotation is available via a time based interval that a user can optionally
+set. Due to the usage of a buffered writer, a size based rotation is not
+possible as the file may not actually get data at each interval.
+
+## Explore Parquet Files
+
+If a user wishes to explore a schema or data in a Parquet file quickly, then
+look at the
```
Comment (DStrand1) [inline plugins/outputs/parquet/README.md:89]: Missing link?

**Review by DStrand1 — COMMENTED** [review]
Preliminary review on the readme, looking over the code now!

**Review by DStrand1 — APPROVED** [review]

File: plugins/outputs/parquet/README.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
+# Parquet Output Plugin
+
+This plugin sends writes metrics to parquet files. By default, the parquet
```
Comment (srebhan) [inline plugins/outputs/parquet/README.md:3]:
```suggestion
This plugin writes metrics to parquet files. By default, the parquet
```

File: plugins/outputs/parquet/parquet.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+func (p *Parquet) Init() error {
+	if p.Directory == "" {
+		p.Directory = "."
+	}
+
+	if p.TimestampFieldName == nil {
+		timestampFieldName := "timestamp"
+		p.TimestampFieldName = &timestampFieldName
+	}
```
Comment (srebhan) [inline plugins/outputs/parquet/parquet.go:53]: Alternatively you could use a non-nil struct member and set the default in the `init()` (lowercase) function allowing an explicit overwrite with an empty string...

File: plugins/outputs/parquet/parquet.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+				rawFields[field.Key] = arrowType
+			}
+		}
+		for _, tag := range metric.TagList() {
+			if _, ok := rawFields[tag.Key]; !ok {
+				arrowType, err := goToArrowType(tag.Value)
+				if err != nil {
+					return nil, fmt.Errorf("error converting '%s=%s' tag to arrow type: %w", tag.Key, tag.Value, err)
+				}
+				rawFields[tag.Key] = arrowType
```
Comment (srebhan) [inline plugins/outputs/parquet/parquet.go:262]: You could directly use a string type here, can't you?

**Review by srebhan — COMMENTED** [review]
The code looks really good @powersj! Just some small comments from my side...

File: plugins/outputs/parquet/README.md (hunk 3 above)
Comment (powersj) [reply plugins/outputs/parquet/README.md:3]: good catch!

File: plugins/outputs/parquet/parquet.go (hunk 5 above)
Comment (powersj) [reply plugins/outputs/parquet/parquet.go:262]: ah yes! thank you

File: plugins/outputs/parquet/parquet.go (hunk 4 above)
Comment (srebhan) [reply plugins/outputs/parquet/parquet.go]: The adaptation is not yet in the code... Did you maybe forget to push that commit?

Comment (powersj) [reply plugins/outputs/parquet/parquet.go]: Are you looking at the "outdated" suggestion, as this whole code block is removed ;)

Comment (srebhan) [reply plugins/outputs/parquet/parquet.go]: Hmmm I don't think so... In the `parquet.go` file line 37 I see
```go
	TimestampFieldName *string         `toml:"timestamp_field_name"`
```

so the `TimestampFieldName` is still a pointer isn't it?

Comment (powersj) [reply plugins/outputs/parquet/parquet.go]: Pushed update to change it to a string, non-pointer

**Review by srebhan — APPROVED** [review]
Thanks @powersj! Code looks fantastic!

**Pull request merged**